go/tools/bzltestutil/lcov.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2022 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bzltestutil
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// ConvertCoverToLcov converts the go coverprofile file coverage.dat.cover to
+// the lcov format and stores it in coverage.dat, where it is picked up by
+// Bazel.
+// The conversion emits line and branch coverage, but not function coverage.
+func ConvertCoverToLcov() error {
+	// The value of test.coverprofile has been set to
+	// ${COVERAGE_OUTPUT_PATH}.cover by the generated TestMain. We have to strip
+	// the ".cover" suffix here so that the generated lcov report is at the
+	// location where Bazel's lcov merge picks it up.
+	inPath := flag.Lookup("test.coverprofile").Value.String()
+	outPath := strings.TrimSuffix(inPath, ".cover")
+
+	in, err := os.Open(inPath)
+	if err != nil {
+		// This can happen if there are no tests and should not be an error.
+		return nil
+	}
+	defer in.Close()
+
+	out, err := os.Create(outPath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	return convertCoverToLcov(bufio.NewScanner(in), out)
+}
+
+var coverLinePattern = regexp.MustCompile(`^(?P<path>.*):(?P<startLine>\d+)\.(?P<startColumn>\d+),(?P<endLine>\d+)\.(?P<endColumn>\d+) (?P<numStmt>\d+) (?P<count>\d+)$`)
+const pathIdx = 1
+const startLineIdx = 2
+const endLineIdx = 4
+const countIdx = 7
+
+func convertCoverToLcov(cover *bufio.Scanner, lcov io.StringWriter) error {
```

Comment:
I added a unit test with a few corner cases I could think of. I don't think that it would make much sense to test failure cases as we generally expect `convertCoverToLcov` to run on valid go cover files. But please let me know if you find that certain test cases are missing.